AdvLoggerPkg/Library/AdvancedLoggerLib/AdvancedLoggerCommon.c

```diff
@@ -69,20 +69,44 @@ AdvancedLoggerMemoryLoggerWrite (
       if ((UsedSize >= LoggerInfo->LogBufferSize) ||
           ((LoggerInfo->LogBufferSize - UsedSize) < EntrySize))
       {
-        //
-        // Update the number of bytes of log that have not been captured
-        //
-        do {
-          CurrentSize = LoggerInfo->DiscardedSize;
-          NewSize     = CurrentSize + (UINT32)NumberOfBytes;
-          OldSize     = InterlockedCompareExchange32 (
-                          (UINT32 *)&LoggerInfo->DiscardedSize,
-                          (UINT32)CurrentSize,
-                          (UINT32)NewSize
-                          );
-        } while (OldSize != CurrentSize);
-
-        return LoggerInfo;
+        if (FeaturePcdGet (PcdAdvancedLoggerAutoWrapEnable) && (LoggerInfo->AtRuntime)) {
+          //
+          // Wrap around the current cursor when auto wrap is enabled on buffer full during runtime.
+          //
+          NewBuffer = LoggerInfo->LogBuffer;
+          OldValue  = InterlockedCompareExchange64 (

```

Comment:
Hmm, as I look at this now, this looks like it will work, but the flow is do CmpExch, reset to the beginning, then do another CmpExch and move LogCurrent to the end of our message. Should we not do this all in one atomic step, check if we are going to be going over the end and if so, set LogCurrent to be LogBuffer + MsgSize?

Reply:
Oh, you mean just do `NewBuffer = LoggerInfo->LogBuffer;` here? That seems better and should work.

Reply:
Hmm.. Simply `CurrentBuffer = LoggerInfo->LogBuffer;` will not work as the `InterlockedCompareExchange64` below will not go through because CurrentBuffer does not match current LoggerInfo->LogCurrent. So either we invoke CmpExch twice, or update the existing `InterlockedCompareExchange64` call below to plum through logic.

Reply:
Actually, I realized we cannot do this, because the read of `LogCurrent` may be stale, so if we always assume we are resetting, someone else may have done it already and we reset and write over valid stuff.